docs/upgrade/0.7.md

```diff
@@ -0,0 +1,122 @@
+| PyQIR 0.6                                     | PyQIR 0.7                         |
+| --------------------------------------------- | --------------------------------- |
+| `pyqir.generator.types.VOID`                  | `pyqir.Type.void(context)`        |
+| `pyqir.generator.types.BOOL`                  | `pyqir.IntType(context, 1)`       |
+| `pyqir.generator.types.Int(width)`            | `pyqir.IntType(context, width)`   |
+| `pyqir.generator.types.DOUBLE`                | `pyqir.Type.double(context)`      |
+| `pyqir.generator.types.QUBIT`                 | `pyqir.qubit_type(context)`       |
+| `pyqir.generator.types.RESULT`                | `pyqir.result_type(context)`      |
+| `pyqir.generator.types.Function(params, ret)` | `pyqir.FunctionType(ret, params)` |
+
+There are two ways to get a `context` object:
+
+1. Use the `context` property on `SimpleModule`.
+   For example:
+
+   ```python
+   from pyqir import SimpleModule, Type
+   module = SimpleModule("name", num_qubits=1, num_results=1)
+   void = Type.void(module.context)
+   ```
+
+2. Create one yourself.
+   But you also need to give the context you created to `SimpleModule`.
+   For example:
+
+   ```python
+   from pyqir import Context, SimpleModule, Type
+   context = Context()
+   module = SimpleModule("name", num_qubits=1, num_results=1, context=context)
+   void = Type.void(context)
+   ```
+
+## Parser
+
+PyQIR 0.7 unified parsing and code generation into a single API that is designed to support both.
+This makes PyQIR much more powerful and will enable workflows that involve inspecting, running passes on, or otherwise transforming QIR that is parsed or generated using PyQIR.
+This means that the API for `pyqir-parser` had to be completely redesigned, which unfortunately makes upgrading challenging.
+Here are some tips.
+
+### Modules
+
+Use `Module.from_bitcode` or `Module.from_ir` instead of the `QirModule` constructor.
+See [](#ir-and-bitcode-conversion).
+
+### Entry points and interop-friendly functions
+
+Instead of `QirModule.entrypoint_funcs`, `QirModule.interop_funcs`, or `QirModule.get_funcs_by_attr`, filter the `Module.functions` list instead.
+For example:
+
+```python
+entry_point = next(filter(pyqir.is_entry_point, module.functions))
+interops = filter(pyqir.is_interop_friendly, module.functions)
+```
+
+### Instructions
+
+The instruction class hierarchy was trimmed down significantly.
+Most subclasses of `QirInstr` were removed.
+The surviving subclasses are `Call`, `FCmp`, `ICmp`, `Phi` and `Switch`.
+
+For any other instruction, it should be possible to use the base `Instruction` class to do anything that was previously possible.
+Use the `opcode` property to check what kind of instruction it is, and the `operands` property to read all of the values that the instruction references.
+The `successors` property is a subset of `operands` and contains just the values that are basic blocks, which can be useful to follow control flow with `br` instructions.
```

Comment:
Using the parser to follow control flow is one of the more complex parts for folks using the parser. If a full example won't fit here, could you point to a specific part of the test folks can draw from, such as: https://github.com/qir-alliance/pyqir/blob/a9d6572165496a7ee02e59ffb0b081f0aa10c9a0/pyqir/tests/test_parser.py#L43-L54 and: https://github.com/qir-alliance/pyqir/blob/a9d6572165496a7ee02e59ffb0b081f0aa10c9a0/pyqir/tests/test_parser.py#L64 to show how to go from a terminator with successors to the corresponding block?

Reply:
To go from a terminator with successors to the corresponding block is just `Instruction.successors[i]`. (The type of `successors` is `List[BasicBlock]`.) I'm not sure what example you had in mind? The test snippets you posted are testing block names which isn't necessary for CFG traversal.

Reply:
Ah, I see where the text says that, but it wasn't immediately clear to me how easy that would be to use. Maybe just an code snippet that shows grabbing a successor and then investigating its terminator would make that more obvious?

Also, I do really like the design here!

Reply:
I added a sentence with an example.